frontend/src/lib/components/Table/utils.ts

```diff
@@ -0,0 +1,215 @@
+                render: filterRender,
+            });
+        }
+    }
+
+    return { dataCells, filterCells, headerCells };
+}
+
+export function recursivelyBuildTableColumnGroups<TData extends Record<string, any>>(
+    tableColumns: TableColumns<TData>,
+    parentSize: number = 100,
+): ColGroupDef[] {
+    const colGroups = [];
+
+    for (const tableColumn of tableColumns) {
+        const columnId = tableColumn.columnId;
+        const columnSize = (tableColumn.sizeInPercent * parentSize) / 100;
+
+        if (isColumnGroupDef(tableColumn)) {
+            const childColGroups = recursivelyBuildTableColumnGroups(tableColumn.subColumns, columnSize);
+
+            // ! The HTML spec doesn't allow nested column-groups, so further col-groups are flattened
+            const flattenedCols = childColGroups.flatMap((colGroup) =>
+                colGroup.cols.map((col) => ({
+                    ...col,
+                    columnId: `${columnId}__${col.columnId}`,
+                })),
+            );
+
+            colGroups.push({
+                cols: flattenedCols,
+                columnId,
+            });
+        } else {
+            // The the table can support dynamic data, so we cannot pre-compute column widths based on data content.
+            // To ensure the table looks somewhat nice, we compute a minimum width that should at least fit the header title and adornments
+
+            let minWidth = getTextWidthWithFont(tableColumn.label, "Equinor", 1);
+            // Padding
+            minWidth += convertRemToPixels(2);
+
+            if (tableColumn.sortable == null || tableColumn.sortable) {
+                // Adornment and gap
+                minWidth += convertRemToPixels(4);
+            }
+
+            colGroups.push({
+                cols: [{ columnId, minWidth, width: columnSize }],
+                columnId,
+            });
+        }
+    }
+
+    return colGroups;
+}
+
+// The table need to at least be wide enough that each column can fit it's headers
+export function computeTableMinWidth(colGroups: ColGroupDef[]) {
+    let minWidth = 0;
+
+    for (const colGroup of colGroups) {
+        for (const col of colGroup.cols) {
+            minWidth += col.minWidth;
+        }
+    }
+
+    return minWidth;
+}
+
+export function useOptInControlledValue<TValue>(
+    initialValue: TValue,
+    controlledProp: TValue | undefined,
+    onValueChange?: (newValue: TValue) => void,
+): [TValue, (newValue: TValue) => void] {
+    const useLocalValue = controlledProp === undefined;
+
+    const [localValue, setLocalValue] = React.useState<TValue>(initialValue);
+
+    const value = useLocalValue ? localValue : controlledProp;
+    const setValue = React.useCallback(
+        function setValue(newValue: TValue) {
+            if (useLocalValue) setLocalValue(newValue);
+            onValueChange?.(newValue);
+        },
+        [useLocalValue, onValueChange],
+    );
+
+    return [value, setValue];
+}
```

Comment:
Minor thought: Should this be placed in a file named `hooks.ts`? As we've usually done that for hooks elsewhere.